### PR TITLE
Add ESMF run-time dependency.

### PR DESCRIPTION
**Pull request by ocefpaf**
Ping @czender it would be nice to add those 3-commands you mentioned in https://github.com/conda-forge/nco-feedstock/pull/5#issuecomment-225577109 as an extra test in this PR. (`conda-build` run any python or bash script named `run_test.{sh,py}` in the test phase.)

### CODE REVIEW

**Comment by ocefpaf** [comment]
We need to sort https://github.com/conda-forge/netcdf-fortran-feedstock/pull/7 first before merging this one.

**Comment by czender** [comment]
@ocefpaf here is the script i recommend to add as run_test.sh. It would be in the NCO feedstock because it tests multiple NCO operators and conservative regridding with an ERWG-generated map. If the last line prints "tst = 1" then everything (NCO and ERWG) works as intended, else something failed. The script below must be modified so that the paths work in the test environment. The file in.nc can be _any_ netCDF file. in.nc is a file that comes with NCO so I used that, but the contents of in.nc do not matter. Everything else is constructed from scratch.

```
ncks -O --rgr skl=~/skl_t42.nc --rgr grid=~/grd_t42.nc --rgr latlon=64,128 --rgr lat_typ=gss --rgr lon_typ=Grn_ctr ~/nco/data/in.nc ~/foo.nc
ncks -O --rgr grid=~/grd_2x2.nc --rgr latlon=90,180 --rgr lat_typ=eqa --rgr lon_typ=Grn_wst ~/nco/data/in.nc ~/foo.nc
ncap2 -O -s 'tst[lat,lon]=1.0f' ~/skl_t42.nc ~/dat_t42.nc
ncremap -a conserve -s ~/grd_t42.nc -g ~/grd_2x2.nc -m ~/map_t42_to_2x2.nc
ncremap -i ~/dat_t42.nc -m ~/map_t42_to_2x2.nc -o ~/dat_2x2.nc
ncwa -O ~/dat_2x2.nc ~/dat_avg.nc
ncks -C -H -v tst ~/dat_avg.nc
```

**Comment by ocefpaf** [comment]
Awesome @czender! That is exactly what we needed!!

I will add that and merge this PR as soon as I fix [netcdf-fortran](https://github.com/conda-forge/netcdf-fortran-feedstock/pull/7) and get a new build of ESMF.

**Comment by ocefpaf** [comment]
@czender I tried your example locally and realized our `nco` build is lacking `ncwa`! Any hints on what we are missing to get that executable in our package?

https://github.com/conda-forge/nco-feedstock/blob/master/recipe/build.sh

**Comment by czender** [comment]
ncwa requires yacc (bison) and maybe flex to build properly. My guess is that those are not in the compile-time environment?

**Comment by ocefpaf** [comment]
Thanks! Let me try adding that `bison`.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -24,20 +24,21 @@ requirements:
         - expat
         - krb5
         - texinfo
+        - bison
+        - flex
+        - m4
```
Comment (ocefpaf) [inline recipe/meta.yaml]: I wonder if `m4` should be a dependency of `flex` and/or `bison`.

File: recipe/run_test.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+# ncks -M http://test.opendap.org:80/opendap/data/ncml/sample_virtual_dataset.ncml
+# ncks -F --dimension samples,1 http://test.opendap.org:80/opendap/data/ncml/sample_virtual_dataset.ncml
+
+ncks -O --rgr skl=skl_t42.nc --rgr grid=grd_t42.nc --rgr latlon=64,128 --rgr lat_typ=gss --rgr lon_typ=Grn_ctr $fin foo.nc
+ncks -O --rgr grid=grd_2x2.nc --rgr latlon=90,180 --rgr lat_typ=eqa --rgr lon_typ=Grn_wst $fin foo.nc
+ncap2 -O -s 'tst[lat,lon]=1.0f' skl_t42.nc dat_t42.nc
+ncremap -a conserve -s grd_t42.nc -g grd_2x2.nc -m map_t42_to_2x2.nc
+ncremap -i dat_t42.nc -m map_t42_to_2x2.nc -o dat_2x2.nc
+ncwa -O dat_2x2.nc dat_avg.nc
+ncks -C -H -v tst dat_avg.nc
```
Comment (ocefpaf) [inline recipe/run_test.sh]: This test worked locally :tada: 

Thanks to @czender conda-forge's `nco` is more robust now. We only need to fix the `netcdf-fortran` build.

**Comment by ocefpaf** [comment]
@bekozi Travis-CI (OS X) is failing because of the library location. Somehow the `nco` [triggered call](https://travis-ci.org/conda-forge/nco-feedstock/builds/137395406#L2350) to `esmf` expects the lib in:

```
dyld: Library not loaded:
/Users/travis/miniconda3/conda-bld/work/esmf/lib/libO/Darwin.gfortran.64.mpiuni.default/libesmf.dylib
```

but the lib is coorectly in `$PREFIX/lib/libesmf.dylib`. Linux is fine though. Any thoughts? Is this something we need to fix in the `esmf` feedstock?

**Comment by bekozi** [comment]
I'll take a look with @rokuingh. I'm not sure why this is only failing on OSX - which is suspicious.

**Comment by bekozi** [comment]
@ocefpaf and @czender this line looks suspicious to me: https://travis-ci.org/conda-forge/nco-feedstock/builds/137395406#L811.

`ENABLE_ESMF...........no`

Could this be related to the issue we are seeing? It is curious that the `esmf-feedstock` links okay with the command line apps on OSX. They are "tested": https://github.com/conda-forge/esmf-feedstock/blob/master/recipe/meta.yaml#L30.

**Comment by ocefpaf** [comment]
> @ocefpaf and @czender this line looks suspicious to me

Nope. That is a future feature for `nco`. Right now it should not need that. (Linux has the same disabled and it works.)

> Could this be related to the issue we are seeing? It is curious that the `esmf-feedstock` links okay with the command line apps on OSX. They are "tested": https://github.com/conda-forge/esmf-feedstock/blob/master/recipe/meta.yaml#L30.

Indeed. I am not familiar with `nco` and `esmf` to debug this, but it may be something simple that we are missing...

**Comment by bekozi** [comment]
Does the actual installation work outside of the conda test step? That is, does it only break during the test step of conda build?

**Comment by czender** [comment]
@ocefpaf is right. esmf is needed only at run-time, not compile time. an alternate hypothesis is that the ncremap executable, which is a script that can do some path mangling, somehow sets paths that are incompatible with the build envrionment.

**Comment by ocefpaf** [comment]
> Does the actual installation work outside of the conda test step? 

Don't know. I don't own a Mac to check that.

> alternate hypothesis is that the `ncremap` executable, which is a script that can do some path mangling, somehow sets paths that are incompatible with the build environment.

I'll take a look. Thanks for the hint @czender.

**Comment by bekozi** [comment]
I'm not really sure where to look next. @czender what is the easiest way to look at `ncremap`'s path mangling?

I doubt this is related, but should there be a `$` before `DYLD` in:

export LIBRARY_SEARCH_VAR=DYLD_FALLBACK_LIBRARY_PATH

https://github.com/conda-forge/nco-feedstock/blob/969e1e90d0e0cf5f32405cc00305bc2a43a1bb24/recipe/build.sh#L4

**Comment by ocefpaf** [comment]
> I doubt this is related, but should there be a `$` before `DYLD` in

That is not it. The `LIBRARY_SEARCH_VAR` is a string holding the proper `LIBRARY_PATH` for Linux or OS X.

**Comment by bekozi** [comment]
There is something fishy going on with the `esmf` OSX build: https://travis-ci.org/NESII/esmf-feedstock/builds/139825295#L422. The related Travis CI script: https://github.com/NESII/esmf-feedstock/blob/eval-conda-install/.travis.yml#L38.

I'm looking into it a bit more and will report back. It's not clear if the issue is happening in the ESMF build or something in the conda build infrastructure.

**Comment by ocefpaf** [comment]
@bekozi there were many changes in conda and conda-build lately and I lost track of them... I don't have the time to investigate this right now, but I will get back to this later on this weekend. Let me known if you find anything.

PS: I do have a PR for `mpich2` almost ready...

**Comment by bekozi** [comment]
> PS: I do have a PR for `mpich2` almost ready...

Sweet.

> Let me known if you find anything.

The linking issue may be on the ESMF Darwin build side. I am going to chat with the build honcho. Basically, the apps (i.e. `ESMF_Info`) are appearing to link to default library locations which is breaking the `rpath` normalization step. Regardless, the `libesmf.dylib` paths are not correct in the apps. See, for example, the `otool` output here: https://travis-ci.org/NESII/esmf-feedstock/builds/139898034#L2368. The path should at least be to the library installed in `${PREFIX}` but the link is to the library in the conda build directory.

Lots of moving parts here. I'll let you know when I know more!

**Comment by bekozi** [comment]
Hi. There is an issue with ESMF apps on Darwin where the dynamic library path is not respecting the ESMF_INSTALL_PREFIX. Working on a fix that I'll plan to introduce as a patch. It may make sense to move forward with the Linux builds in the meantime.

**Comment by ocefpaf** [comment]
Thanks to https://github.com/conda-forge/esmf-feedstock/pull/14 this is now passing! :tada:

**Pull request merged**